artemis-server/src/main/java/org/apache/activemq/artemis/spi/core/security/jaas/SCRAMPropertiesLoginModule.java

```diff
@@ -0,0 +1,237 @@
+import javax.security.auth.login.LoginException;
+
+import org.apache.activemq.artemis.spi.core.security.scram.SCRAM;
+import org.apache.activemq.artemis.spi.core.security.scram.ScramException;
+import org.apache.activemq.artemis.spi.core.security.scram.ScramUtils;
+import org.apache.activemq.artemis.spi.core.security.scram.StringPrep;
+import org.apache.activemq.artemis.spi.core.security.scram.StringPrep.StringPrepError;
+import org.apache.activemq.artemis.spi.core.security.scram.UserData;
+import org.apache.activemq.artemis.utils.PasswordMaskingUtil;
+
+/**
+ * Login modules that uses properties files similar to the {@link PropertiesLoginModule}. It can
+ * either store the username-password in plain text or in an encrypted/hashed form. the
+ * {@link #main(String[])} method provides a way to prepare unencrypted data to be encrypted/hashed.
+ */
+public class SCRAMPropertiesLoginModule extends PropertiesLoader implements AuditLoginModule {
+
+   /**
+    *
+    */
+   private static final String SEPARATOR_MECHANISM = "|";
+   private static final String SEPARATOR_PARAMETER = ":";
+   private static final int MIN_ITERATIONS = 4096;
+   private static final SecureRandom RANDOM_GENERATOR = new SecureRandom();
+   private Subject subject;
+   private CallbackHandler callbackHandler;
+   private Properties users;
+   private Map<String, Set<String>> roles;
+   private UserData userData;
+   private String user;
+   private final Set<Principal> principals = new HashSet<>();
+
+   @Override
+   public void initialize(Subject subject, CallbackHandler callbackHandler, Map<String, ?> sharedState,
+                          Map<String, ?> options) {
+      this.subject = subject;
+      this.callbackHandler = callbackHandler;
+
+      init(options);
+      users = load(PropertiesLoginModule.USER_FILE_PROP_NAME, "user", options).getProps();
+      roles = load(PropertiesLoginModule.ROLE_FILE_PROP_NAME, "role", options).invertedPropertiesValuesMap();
+
+   }
+
+   @Override
+   public boolean login() throws LoginException {
+      NameCallback nameCallback = new NameCallback("Username: ");
+      executeCallbacks(nameCallback);
+      user = nameCallback.getName();
+      SCRAMMechanismCallback mechanismCallback = new SCRAMMechanismCallback();
+      executeCallbacks(mechanismCallback);
+      SCRAM scram = getTypeByString(mechanismCallback.getMechanism());
+      if (user == null) {
+         userData = generateUserData(null); // generate random user data
+      } else {
+         String password = users.getProperty(user + SEPARATOR_MECHANISM + scram.name());
+         if (password == null) {
+            // fallback for probably unencoded user/password or a single encoded entry
+            password = users.getProperty(user);
+         }
+         if (PasswordMaskingUtil.isEncMasked(password)) {
+            String[] unwrap = PasswordMaskingUtil.unwrap(password).split(SEPARATOR_PARAMETER);
```

Comment:
Probably worth verifying this has the exact number of expected elements, to give a better failure if it doesnt?

Comment trigged by memory of wondering how this stuff interacts with the the cli commands for adding users.

Reply:
I think we can defer cli support, the main provides a usable mechanism to register users. There may be a need to have some sort of plugable user registration thing for the cli, but that can be explored in the future

Reply:
Sorry I don't understand this here. if the properties file is messed then we are out of luck and this needs to be fixed, the implementation is not a debugging util. I also don't get what this has to do with user registration?